File: src/main/java/dev/stashy/extrasounds/SoundManager.java
package dev.stashy.extrasounds;

import dev.stashy.extrasounds.debug.DebugUtils;
import dev.stashy.extrasounds.sounds.SoundType;
import dev.stashy.extrasounds.sounds.Sounds;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.sound.PositionedSoundInstance;
import net.minecraft.client.sound.SoundInstance;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.item.ItemStack;
import net.minecraft.sound.SoundCategory;
import net.minecraft.sound.SoundEvent;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.registry.Registry;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class SoundManager
{
    private static final Logger LOGGER = LogManager.getLogger();
    private static long lastPlayed = System.currentTimeMillis();

    public static void playSound(ItemStack stack, SoundType type)
    {
        var itemId = Registry.ITEM.getId(stack.getItem());
        String idString = ExtraSounds.getClickId(itemId, type);
        if (!Identifier.isValid(idString))
        {
            LOGGER.error("Unable to parse sound from ID: " + idString);
            return;
        }

        Identifier id = Identifier.tryParse(idString);
        Registry.SOUND_EVENT.getOrEmpty(id).ifPresentOrElse(
                (snd) -> playSound(snd, type),
                () -> LOGGER.error("Sound cannot be found in registry: " + id));
    }

    public static void playSound(StatusEffect effect, boolean add)
    {
        DebugUtils.effectLog(effect, add);

        SoundEvent e = add ?
                switch (effect.getCategory())
                        {
                            case HARMFUL -> Sounds.EFFECT_ADD_NEGATIVE;
                            case NEUTRAL, BENEFICIAL -> Sounds.EFFECT_ADD_POSITIVE;
                        }
                :
                switch (effect.getCategory())
                        {
                            case HARMFUL -> Sounds.EFFECT_REMOVE_NEGATIVE;
                            case NEUTRAL, BENEFICIAL -> Sounds.EFFECT_REMOVE_POSITIVE;
                        };
        playSound(e, SoundType.EFFECT);
    }

    public static void playSound(SoundEvent snd, SoundType type)
    {
        playSound(snd, type, type.category);
    }

    public static void playSound(SoundEvent snd, SoundType type, SoundCategory cat)
    {
        playSound(snd, type.pitch, cat);
    }

    public static void playSound(SoundEvent snd, float pitch, SoundCategory cat)
    {
        playSound(new PositionedSoundInstance(snd.getId(), cat, getMasterVol(), pitch, false, 0,
                                              SoundInstance.AttenuationType.NONE, 0.0D, 0.0D, 0.0D,
                                              true));
        DebugUtils.soundLog(snd);
    }

    public static void playSound(SoundEvent snd, SoundType type, BlockPos position)
    {
        playSound(new PositionedSoundInstance(snd, type.category, getMasterVol(), type.pitch,
                                              position.getX() + 0.5,
                                              position.getY() + 0.5,
                                              position.getZ() + 0.5));
        DebugUtils.soundLog(snd);
    }

    public static void playSound(PositionedSoundInstance instance)
    {
        throttle(() -> {
            var client = MinecraftClient.getInstance();
            client.send(() -> {
                client.getSoundManager().play(instance);
            });
        });
    }

    public static void stopSound(SoundEvent e, SoundType type)
    {
        MinecraftClient.getInstance().getSoundManager().stopSounds(e.getId(), type.category);
    }

    private static void throttle(Runnable r)
    {
        try
        {
            long now = System.currentTimeMillis();
            if (now - lastPlayed > 5) r.run();
            lastPlayed = now;
        }
        catch (Exception e)
        {
            System.err.println("Failed to play sound:");
            e.printStackTrace();
        }
    }

    private static float getMasterVol()
    {
        return MinecraftClient.getInstance().options.getSoundVolume(Mixers.MASTER);
    }
}

File: src/main/kotlin/dev/stashy/extrasounds/SoundManager.kt
package dev.stashy.extrasounds

import dev.stashy.extrasounds.Main.Companion.getClickId
import dev.stashy.extrasounds.sounds.SoundType
import dev.stashy.extrasounds.sounds.Sounds
import net.minecraft.client.MinecraftClient
import net.minecraft.client.sound.PositionedSoundInstance
import net.minecraft.client.sound.SoundInstance
import net.minecraft.item.Item
import net.minecraft.sound.SoundCategory
import net.minecraft.sound.SoundEvent
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import org.apache.logging.log4j.LogManager

object SoundManager {
    private var lastPlayed = 0L

    fun playItemSound(item: Item, type: SoundType, interval: Long = 5) {
        playSound(getItemSound(item, type), type, interval = interval)
    }

    fun getItemSound(item: Item, type: SoundType): SoundEvent {
        val id = getClickId(Registry.ITEM.getId(item), type)
        if (!Identifier.isValid(id)) {
            LogManager.getLogger().warn("Sound ID is not valid: \"$id\", returning default")
            return Sounds.ITEM_PICK
        }
        return SoundEvent(Identifier.tryParse(id))
    }

    @JvmOverloads
    fun playSound(
        snd: SoundEvent,
        type: SoundType,
        pitch: Float = type.pitch,
        category: SoundCategory = type.category,
        pos: Triple<Double, Double, Double> = Triple(0.0, 0.0, 0.0),
        relative: Boolean = true,
        interval: Long = 5,
    ) {
        playSound(
            PositionedSoundInstance(
                snd.id, category, getMasterVol(), pitch,
                SoundInstance.createRandom(), false, 0,
                SoundInstance.AttenuationType.NONE, pos.first, pos.second, pos.third, relative
            ), interval
        )
    }

    fun playSound(sound: SoundInstance, interval: Long = 5) {
        throttle({
            MinecraftClient.getInstance().apply {
                send { soundManager.play(sound) }
            }
        }, interval)
    }

    fun stopSound(evt: SoundEvent, type: SoundType) {
        MinecraftClient.getInstance().soundManager.stopSounds(evt.id, type.category)
    }

    fun throttle(fn: () -> Unit, interval: Long = 5) {
        val now = System.currentTimeMillis()
        if (now - lastPlayed >= interval) {
            fn()
            lastPlayed = now
        }
    }

    private fun getMasterVol(): Float =
        MinecraftClient.getInstance().options.getSoundVolume(Mixers.Master)

}